oshmem/shmem/c/shmem_mcs_lock.c

```diff
@@ -0,0 +1,300 @@
+        RUNTIME_CHECK_RC(rc);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	if (next_value == NEXT_MASK) {
+		swap_cond = SET_SIGNAL(my_pe);
+
+		/**
+		 * prev_value = shmem_atomic_compare_swap(tail, swap_cond, 0,
+		 *					   mcs_tail_owner);
+		 */
+		rc = MCA_ATOMIC_CALL(cswap(oshmem_ctx_default,
+					   (void *)tail,
+					   (uint64_t *)&(prev_value),
+					   OSHMEM_ATOMIC_PTR_2_INT(&swap_cond,
+						sizeof(swap_cond)),
+					   OSHMEM_ATOMIC_PTR_2_INT(&val_tmp,
+						sizeof(val_tmp)),
+					   sizeof(int),
+					   mcs_tail_owner));
+		RUNTIME_CHECK_RC(rc);
+
+		/** I am the tail.. and lock is released */
+		if (prev_value == swap_cond)
+			return;
+		/**
+		 * I am not the tail, another PE maybe racing to acquire lock,
+		 * let them complete setting themselves as our next
+		 */
+		nmask = NEXT_MASK;
+		while(next_value == nmask) {
+			/**
+			 * shmem_wait_until(next, SHMEM_CMP_NE, NEXT_MASK);
+			 * next_value = shmem_atomic_fetch(next, my_pe);
+			 **/
+
+			rc = MCA_SPML_CALL(wait((void*)next, SHMEM_CMP_NE,
+						(void*)&nmask, SHMEM_INT));
+			RUNTIME_CHECK_RC(rc);
+			rc = MCA_ATOMIC_CALL(fadd(oshmem_ctx_default,
+						  (void*)next,
+						  (void*)&next_value,
+						  tval, sizeof(int),
+						  my_pe));
+			RUNTIME_CHECK_RC(rc);
+		}
+	}
+	/** There is a successor release lock to the successor
+	    shmem_atomic_add(next, SIGNAL_MASK, NEXT(next_value));
+	    shmem_quiet();
+	*/
+	a_val = SIGNAL_MASK;
+	rc = MCA_ATOMIC_CALL(add(oshmem_ctx_default,
+				 (void *)next, a_val,
+				 sizeof(a_val), NEXT(next_value)));
+	RUNTIME_CHECK_RC(rc);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+}
+
+static inline int
+shmem_mcs_internal_test_lock(long *lockp)
+{
+	lock_t	*lock			= (lock_t *) lockp;
+	int	mcs_tail_owner		= MCSQ_TAIL_OWNER(lock);
+	int	*tail			= &(lock->tail);
+	int	*next			= &(lock->next);
+	int	my_pe			= shmem_my_pe();
+	int	new_tail_req		= 0;
+	int	prev_tail		= 0;
+	int	tmp_cond		= 0;
+	int	rc			= 0;
+
+	/** Initializing next pointer to next mask */
+	*next = NEXT_MASK;
+
+	/** Signal for setting lock */
+	new_tail_req = SET_SIGNAL(my_pe);
+
+	/**
+	 * Check if previously cleared before swapping
+	 * prev_tail	 = shmem_atomic_compare_swap(tail, 0,
+	 *					     new_tail_req,
+	 *					     mcs_tail_owner);
+	 */
+	rc = MCA_ATOMIC_CALL(cswap(oshmem_ctx_default,
+				   (void *)tail,
+				   (uint64_t *)&(prev_tail),
+				   OSHMEM_ATOMIC_PTR_2_INT(&tmp_cond,
+					sizeof(tmp_cond)),
+				   OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req,
+					sizeof(new_tail_req)),
+				   sizeof(int), mcs_tail_owner));
+	RUNTIME_CHECK_RC(rc);
+
+	if (0 == prev_tail)
+		return 0; /** lock acquired successfully */
+	return 1; /** lock is set but returning without waiting */
+}
+
+void _shmem_mcs_set_lock(long *lock) {
+	shmem_mcs_internal_set_lock(lock);
+}
+
+void _shmem_mcs_clear_lock(long *lock) {
+	shmem_mcs_internal_clear_lock(lock);
+}
+
+int  _shmem_mcs_test_lock(long *lock) {
+	return shmem_mcs_internal_test_lock(lock);
```

Comment:
why do we need these wrappers?

Reply:
Was intended to keep similar to the existing implementation. Doesn't have purpose for this implementation. Removing it.